### ISSUE TITLE
Unexpected error with numpy==2.4.0rc1 when specifying `--follow-imports`

### ISSUE THREAD

**Issue by TomAugspurger** [python/mypy#20364]
**Bug Report**

I see an error when type checking a code base that specifies`--follow-imports=skip/error`,  even when the dependency is importable. The error is not reproducible with `--follow-imports=normal/silent`.

**To Reproduce**

```python
# file: bug.py
import numpy as np

np_dtype: np.dtype[np.object_] = np.dtype("object")
```

Run mypy **with numpy==2.4.0rc1** in the environment. Using `uv run`:

```python
uv run --python=3.13 --with numpy==2.4.0rc1 --with mypy mypy bug.py --follow-imports=skip
```

**Expected Behavior**

No error.

**Actual Behavior**

```
bug.py:3: error: Incompatible types in assignment (expression has type "dtype[float64]", variable has type "dtype[object_]")  [assignment]
Found 1 error in 1 file (checked 1 source file)
```

Strangely, I get the same behavior with `--follow-improts=error`, which makes me think this is an issue in mypy, though it could still be something in numpy given that it's not reproducible with numpy==2.3.5:

```
$ uv run --python=3.13 --with numpy==2.4.0rc1 --with mypy mypy bug.py --follow-imports=error
bug.py:3: error: Incompatible types in assignment (expression has type "dtype[float64]", variable has type "dtype[object_]")  [assignment]
```

**Your Environment**
- 
- Mypy version used: 1.19.0
- Mypy command-line flags: `--follow-imports=skip/error`
- Mypy configuration options from `mypy.ini` (and other config files): none
- Python version used: 3.13

xref https://github.com/rapidsai/cudf/issues/20782#issuecomment-3614041785.

**Comment by jorenham** [comment]
It might be related to a change in the first `numpy.dtype.__new__` overload with `-> dtype[float64]` that now also accepts `_Float64Codes | _DoubleCodes` (defined in [`numpy._typing._char_codes`](https://github.com/numpy/numpy/blob/main/numpy/_typing/_char_codes.py)), which used to be in a separate overload further down. Because if mypy somehow doesn't know what that type is, then it'll effectively become `Any`, and that first overload will always be matched.
If that's indeed the case, then the question becomes: Why isn't mypy able to infer those types anymore when `--follow-imports` is used?

**Comment by hauntsaninja** [comment]
Note if you use `--follow-imports=skip/error` you should basically not expect types from imported libraries to work well....

(I've tried to heavily recommend against using these in the docs, do you have a reson to do so?)

(The answer in this case is probably that something is in a .py file, not a .pyi file. mypy does not use the follow imports setting for stubs, unless you also set follow imports for stubs)

**Comment by jorenham** [comment]
> (The answer in this case is probably that something is in a .py file, not a .pyi file. mypy does not use the follow imports setting for stubs, unless you also set follow imports for stubs)

Ah that must be it then, because `numpy._typing._char_codes` is indeed defined in a `_char_codes.py` and `numpy.dtype` in `__init__.pyi`.

**Comment by TomAugspurger** [comment]
> (I've tried to heavily recommend against using these in the docs, do you have a reson to do so?)

Yep, the docs were very clear on that, so we're taking this as an opportunity to update our configuration. I'll close this then, unless you think there's something else to document here.
